### PR TITLE
Adds CRC deployment automation and CI documentation

### PR DESCRIPTION
**Pull request by perdasilva**
This PR:
* Introduces new scripts to provision a crc cluster and deploy olm onto it
* Introduces a new `values-crc-e2e.yaml` to adjust the images, arguments, and ImagePullPolicy of the deployments
* Introduces additional `*.crc.e2e.yaml` yaml patch files to patch additional olm manifests
* Adds new make targets for interacting with th deployment scripts
* Adds documentation on how ci and downstream sync work
* Adds documentation on how to use crc for e2e tests

### CODE REVIEW

File: docs/local-testing-with-crc.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# Local testing with CRC
+
+We can use CRC as an Openshift-like environment within which to test downstream OLM. [CRC](https://developers.redhat.com/products/codeready-containers/overview) 
+is a tool for deploying a local Openshift cluster on your laptop.
+
+TL;DR
+
+1. Install [CRC](https://developers.redhat.com/products/codeready-containers/overview)
+2. `make crc` to provision a CRC cluster, build OLM, and deploy it on the cluster
+3. `make e2e WHAT=operator-lifecycle-manager E2E_INSTALL_NS=openshift-operator-lifecycle-manager E2E_TEST_NS=openshift-operators`, i.e., execute e2e tests as you normally would.
```
Comment (timflannagan) [inline docs/local-testing-with-crc.md:10]: This command looks like a mirror of what `make e2e/olm` is doing under-the-hood. It just shells out to OLM's `make e2e` command and overrides the install/test namespace variables to point to the downstream ones.

File: scripts/crc-deploy.sh [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+# This script manages the deployment of downstream olm on a local CRC cluster
+# It will deploy the local images: olm:test and opm:test
+# Both built with make build/olm-container and build/registry-container respectively
+
+echo "Deploying OLM to CRC"
+
+# Set kubeconfig to CRC if necessary
+export KUBECONFIG=${KUBECONFIG:-${HOME}/.crc/machines/crc/kubeconfig}
+KUBE_ADMIN_PASSWORD=$(tr -d '\n' < "${HOME}/.crc/machines/crc/kubeadmin-password")
```
Comment (timflannagan) [inline scripts/crc-deploy.sh:13]: IIRC there's a dedicated `crc console --credentials` command that might be better to parse?

This is what I have locally as an alias:

```console
$ alias crc_l
crc_l='eval $(crc console --credentials | cut -d" -f2- | grep -v developer | tr -d ")'
```

File: scripts/crc-deploy.sh [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+# Push olm image
+echo "Pushing olm image"
+docker tag olm:test localhost:5000/openshift/olm:test
+docker push localhost:5000/openshift/olm:test
+
+# Push registry image
+echo "Pushing registry image"
+docker tag opm:test localhost:5000/openshift/opm:test
+docker push localhost:5000/openshift/opm:test
```
Comment (timflannagan) [inline scripts/crc-deploy.sh:78]: Maybe this can be thrown in a function if we're trapping signals and require cleanup?

Comment (timflannagan) [reply scripts/crc-deploy.sh:78]: Also variable to track any of the opm/olm test images?

File: scripts/sync_pop_candidate.sh [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+    # 1. Pop next commit off cherrypick set
+    # 2. Cherry-pick
+    # 3. Ammend commit
+    # 4. Remove from cherrypick set
+    make vendor
+    make manifests
+    git add .
+    git status
+    git commit --amend --allow-empty --no-edit --trailer "Upstream-repository: ${remote}" --trailer "Upstream-commit: ${rc}"
```
Comment (timflannagan) [inline scripts/sync_pop_candidate.sh:63]: Note: the --trailer command requires new git versions. I saw locally that I needed to build git from source as f34 (f35?) didn't have this in the dnf stack.

File: docs/local-testing-with-crc.md (hunk 1 above)
Comment (perdasilva) [reply docs/local-testing-with-crc.md:10]: Nice, fixed!

File: scripts/crc-deploy.sh (hunk 2 above)
Comment (perdasilva) [reply scripts/crc-deploy.sh:13]: Oh, found an even nicer way: `crc console --credentials -o json | jq -r .clusterConfig.adminCredentials.password`

File: scripts/crc-deploy.sh (hunk 3 above)
Comment (perdasilva) [reply scripts/crc-deploy.sh:78]: Fixed

File: scripts/sync_pop_candidate.sh (hunk 4 above)
Comment (perdasilva) [reply scripts/sync_pop_candidate.sh:63]: Do you happen to know from which version --trailer was introduced? I'm having trouble finding it. I could put in a check and a link.

Comment (timflannagan) [reply scripts/sync_pop_candidate.sh:63]: IIRC it's git 2.33

File: scripts/crc-deploy.sh [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+  OPM_IMG="$(oc get istag/opm:latest -o json | jq -r .image.dockerImageReference)"
+
+  make_manifest_patches "${OLM_IMG}" "${OPM_IMG}"
+
+  # Build e2e manifests
+  echo "Generating manifests"
+  # CRC_E2E_VALUES is already set in this script and exported
+  ./scripts/generate_crds_manifests.sh
+
+  # Apply patches
```
Comment (dinhxuanvu) [inline scripts/crc-deploy.sh:154]: I think we should apply the OLM manifests first and then apply the psm and profiler deployments.

File: Makefile [hunk 6]
```diff
@@ -162,6 +166,27 @@ verify:
...
+crc-start:
+	echo "Starting CRC"
+	./scripts/crc-start.sh
+
+.PHONY: crc-build
+crc-build:
+	echo "Building olm image"
+	IMG="olm:test" $(MAKE) build/olm-container
+	echo "Building opm image"
+	rm -rf bin
```
Comment (timflannagan) [inline Makefile:179]: Any idea on why this rm command was needed between build steps?

File: scripts/crc-deploy.sh (hunk 5 above)
Comment (timflannagan) [reply scripts/crc-deploy.sh:154]: Interestingly enough, it looks like do the exact opposite in the context of CVO run levels looking at our root manifests directory: https://github.com/openshift/operator-framework-olm/tree/master/manifests.

Comment (dinhxuanvu) [reply scripts/crc-deploy.sh:154]: 00 is higher priority right?

File: Makefile (hunk 6 above)
Comment (perdasilva) [reply Makefile:175]: I'll add a comment. But basically, if we don't rm bin before building the opm container and the binaries are still in bin, it won't (re-)compile them and you'll end up with a binary that doesn't work inside the container =S

File: scripts/crc-deploy.sh (hunk 5 above)
Comment (perdasilva) [reply scripts/crc-deploy.sh:154]: @dinhxuanvu that's right. I'm updating the script to pipe to sort. Then it should apply everything in priority order. Good catch! The comment here was stale. What I was doing before was deleting them in reverse priority order then applying them in priority order. But then I discovered replace.

File: Makefile (hunk 6 above)
Comment (perdasilva) [reply Makefile]: I'm cleaning this up by adding a clean target and calling it before the build/*-container targets

File: scripts/crc-deploy.sh [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+export CRC_E2E_VALUES="values-crc-e2e.yaml"
+
+# Set kubeconfig to CRC if necessary
+export KUBECONFIG=${KUBECONFIG:-${HOME}/.crc/machines/crc/kubeconfig}
+KUBE_ADMIN_USER=$(crc console --credentials -o json | jq -r .clusterConfig.adminCredentials.username)
+KUBE_ADMIN_PASSWORD=$(crc console --credentials -o json | jq -r .clusterConfig.adminCredentials.password)
+
+# login to crc
+echo "Logging in as kubeadmin"
+oc login -u "${KUBE_ADMIN_USER}" -p "${KUBE_ADMIN_PASSWORD}" > /dev/null
```
Comment (timflannagan) [inline scripts/crc-deploy.sh:132]: I just tried pulling down these changes, running locally, and I got hit with the following error:

```console
$ make crc
...
echo "Deploying OLM"
Deploying OLM
./scripts/crc-deploy.sh
Deploying OLM to CRC
Logging in as kubeadmin
error: couldn't get https://127.0.0.1:42973/.well-known/oauth-authorization-server: unexpected response status 403
make: *** [Makefile:180: crc-deploy] Error 1
```

Going to try and manually run the crc-deploy Makefile target and see whether that authentication issue is reproducible.

Comment (timflannagan) [reply scripts/crc-deploy.sh:132]: Okay I figured out the issue: in the case where a user already has a defined $KUBECONFIG variable locally, and that kubeconfig contains many contexts', we're not changing the default context to the crc instance:

```console
$ echo $KUBECONFIG
/home/tflannag/.kube/config
$ export KUBECONFIG=${KUBECONFIG:-${HOME}/.crc/machines/crc/kubeconfig}
$ echo $KUBECONFIG
/home/tflannag/.kube/config
$
$ KUBE_ADMIN_USER=$(crc console --credentials -o json | jq -r .clusterConfig.adminCredentials.username)
$ KUBE_ADMIN_PASSWORD=$(crc console --credentials -o json | jq -r .clusterConfig.adminCredentials.password)
$ oc login -u "${KUBE_ADMIN_USER}" -p "${KUBE_ADMIN_PASSWORD}" > /dev/null
Logging in as kubeadmin
error: couldn't get https://127.0.0.1:42973/.well-known/oauth-authorization-server: unexpected response status 403
$ crc status
CRC VM:          Running
OpenShift:       Running (v4.10.3)
Podman:          
Disk Usage:      19.83GB of 32.74GB (Inside the CRC VM)
Cache Usage:     103.6GB
Cache Directory: /home/tflannag/.crc/cache
$ kubectx -c
kind-kind
$ # switch to the crc-admin context
$ kubectx crc-admin
$ oc login -u "${KUBE_ADMIN_USER}" -p "${KUBE_ADMIN_PASSWORD}" > /dev/null
$ echo $?
0
```

File: scripts/crc-deploy.sh [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+  # Build e2e manifests
+  echo "Generating manifests"
+  # CRC_E2E_VALUES is already set in this script and exported
+  # If set, it will include the file referenced by it in the helm template command
+  # and update the manifests to use the locally built images
+  ./scripts/generate_crds_manifests.sh
+
+  # Apply patches
+  ${YQ} write --inplace -s scripts/psm-operator-deployment.crc.e2e.patch.yaml manifests/0000_50_olm_06-psm-operator.deployment.yaml
+  ${YQ} write --inplace -s scripts/collect-profiles.crc.e2e.patch.yaml manifests/0000_50_olm_07-collect-profiles.cronjob.yaml
```
Comment (timflannagan) [inline scripts/crc-deploy.sh:162]: Ran into another error while running this locally and manually setting the current kubecontext to crc-admin:

```console
$ make crc-deploy
...
wrote ./tmp.hxG09zV0mc/chart/olm/templates/0000_50_olm_13-operatorgroup-default.yaml
wrote ./tmp.hxG09zV0mc/chart/olm/templates/0000_50_olm_13-operatorgroup-default.yaml
wrote ./tmp.hxG09zV0mc/chart/olm/templates/0000_90_olm_01-prometheus-rule.yaml
wrote ./tmp.hxG09zV0mc/chart/olm/templates/0000_90_olm_00-service-monitor.yaml
wrote ./tmp.hxG09zV0mc/chart/olm/templates/0000_90_olm_00-service-monitor.yaml

go run: no packages loaded from ./vendor/github.com/mikefarah/yq/v3/
make: *** [Makefile:180: crc-deploy] Error 1
```

File: scripts/crc-deploy.sh [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+- command: update
+  path: spec.jobTemplate.spec.template.spec.containers[0].image
+  value: OLM_IMAGE
+EOF
+
+  sed "${SED_OPTS[@]}" "${COLLECT_PROFILE_PATCH_TEMPLATE}" > scripts/collect-profiles.crc.e2e.patch.yaml
+}
+
+# YQ for applying yaml patches
+YQ="go run ./vendor/github.com/mikefarah/yq/v3/"
```
Comment (timflannagan) [inline scripts/crc-deploy.sh:119]: See my earlier comment about running into issues when getting to the part in this workflow where we use yq. I tried running the following command locally thinking we might have support for overriding YQ:

```console
$ make crc-deploy YQ=/usr/local/bin/yq
...
go run: no packages loaded from ./vendor/github.com/mikefarah/yq/v3/
```

Maybe this is something we can override just like the KUBECONFIG environment variable?

Comment (timflannagan) [reply scripts/crc-deploy.sh:119]: Andddd I tried manually overriding this in the crc-deploy.sh bash script and realized I have a v4 version locally:

```console
$ make crc-deploy YQ=/usr/local/bin/yq
...
Error: write inplace cannot be used with split file
make: *** [Makefile:180: crc-deploy] Error 1
$ yq --version
yq (https://github.com/mikefarah/yq/) version 4.23.1
```

Comment (timflannagan) [reply scripts/crc-deploy.sh:119]: It looks like we're stranding some changes made while generating the CRC patches:

```console
$ git status
	modified:   manifests/0000_50_olm_07-olm-operator.deployment.ibm-cloud-managed.yaml
	modified:   manifests/0000_50_olm_07-olm-operator.deployment.yaml
	modified:   manifests/0000_50_olm_08-catalog-operator.deployment.ibm-cloud-managed.yaml
	modified:   manifests/0000_50_olm_08-catalog-operator.deployment.yaml
	modified:   pkg/manifests/csv.yaml
```

Maybe we need to wrap this in it's own function so we can enact a cleanup function (e.g. `git restore manifests`) when we run into a non-zero exit code?

**Comment by perdasilva** [comment]
/override e2e-gcp
/override e2e-gcp-console-olm
/override e2e-upgrade

Overriding as these changes don't affect anything in the product or pipeline

**Pull request merged**